### PR TITLE
Fix Walloc-size-larger-than warnings in SimG4CMS/Forward/*NumberingScheme

### PR DESCRIPTION
**Pull request by iarspider**
#### PR description:

Fix warnings reported here: https://github.com/cms-sw/cmssw/issues/41795 . The changes are similar to what was done in https://github.com/cms-sw/cmssw/pull/41826 .

#### PR validation:

No special validation

### CODE REVIEW

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test for CMSSW_13_2_DBG_X

just in case - some code changes are conditionally compiled only in DBG_X

**Comment by civanch** [comment]
@iarspider , I fully agree that making temporary C-arrays in these classes is not necessary, in general, this PR is valid. However, these classes are used by sub-detectors, which in normal tests are not involved, so validation of results should be done with enough statistics or in the new code you should have another proof that the result is not changed. In a standard WF you likely will not see anything. A safe approach would be to do exactly the same thing as before but without C-arrays.

File: SimG4CMS/Forward/src/BHMNumberingScheme.cc [hunk 1]
```diff
@@ -6,43 +6,20 @@
...
-  int level = detectorLevel(aStep);
 
-  LogDebug("BHMSim") << "BHMNumberingScheme number of levels= " << level;
-  if (level > 0) {
-    int* copyno = new int[level];
-    G4String* name = new G4String[level];
-    detectorLevel(aStep, level, copyno, name);
+  //Find number of levels
+  const G4VTouchable* touch = aStep->GetPreStepPoint()->GetTouchable();
+  int level = (touch) ? ((touch->GetHistoryDepth()) + 1) : 0;
```
Comment (civanch) [inline SimG4CMS/Forward/src/BHMNumberingScheme.cc:14]: I do not think that touch can be undefined.

Comment (iarspider) [reply SimG4CMS/Forward/src/BHMNumberingScheme.cc:14]: The check was there in the [original code](https://github.com/cms-sw/cmssw/pull/41942/files/3c866ea649c4da1c3dc5ba483057eb7307f5c1cf#diff-5b5049e9d92e98458c2acbd42d3ea28d19afb6062dd73ea2629c3e68530c37f8L12), I can drop it if you think it's safe.

File: SimG4CMS/Forward/src/BHMNumberingScheme.cc [hunk 2]
```diff
@@ -6,43 +6,20 @@
...
-      int subdet = copyno[0];
-      int zside = copyno[3];
-      int station = copyno[1];
-      intindex = packIndex(subdet, zside, station);
-      LogDebug("BHMSim") << "BHMNumberingScheme : subdet " << subdet << " zside " << zside << " station " << station;
-    }
-    delete[] copyno;
-    delete[] name;
+  LogDebug("BHMSim") << "BHMNumberingScheme number of levels= " << level;
+  if (level > 3) {
```
Comment (civanch) [inline SimG4CMS/Forward/src/BHMNumberingScheme.cc:17]: In previous implementation if(level>3) the fix depth is used not depended on level of 'level'

File: SimG4CMS/Forward/src/BscNumberingScheme.cc [hunk 3]
```diff
@@ -13,69 +13,51 @@
...
+                         << "ich=" << ich << "copyno" << copyno << "name=" << name;
     }
     intindex = packBscIndex(zside, det, station);
     LogDebug("BscSim") << "BscNumberingScheme : det " << det << " zside " << zside << " station " << station
                        << " UnitID 0x" << std::hex << intindex << std::dec;
 
-    for (int ich = 0; ich < level; ich++)
-      LogDebug("BscSim") << " name = " << name[ich] << " copy = " << copyno[ich];
-
+    for (int ich = 0; ich < level; ich++) {
```
Comment (civanch) [inline SimG4CMS/Forward/src/BscNumberingScheme.cc:55]: Why we need this loop? LogDebug may be moved inside previous loop after line 31

File: SimG4CMS/Forward/src/ZdcNumberingScheme.cc [hunk 4]
```diff
@@ -107,16 +107,16 @@ unsigned int ZdcNumberingScheme::getUnitID(const G4Step* aStep) const {
 
     edm::LogVerbatim("ForwardSim") << "ZdcNumberingScheme:"
                                    << "  getUnitID - # of levels = " << level;
-    for (int ich = 0; ich < level; ich++)
-      edm::LogVerbatim("ForwardSim") << "  " << ich << ": copyno " << copyno[ich] << " name=" << name[ich]
-                                     << "  section " << section << " zside " << zside << " layer " << layer << " fiber "
-                                     << fiber << " channel " << channel << "packedIndex =" << intindex
-                                     << " detId raw: " << std::hex << index << std::dec;
+    for (int ich = 0; ich < level; ich++) {
```
Comment (civanch) [inline SimG4CMS/Forward/src/ZdcNumberingScheme.cc:110]: This loop is also not needed, LogVerbatim may be moved after the line 45

File: SimG4CMS/Forward/src/BHMNumberingScheme.cc (hunk 2 above)
Comment (iarspider) [reply SimG4CMS/Forward/src/BHMNumberingScheme.cc:17]: Not sure what you mean. In [previous implementation](https://github.com/cms-sw/cmssw/blob/master/SimG4CMS/Forward/src/BHMNumberingScheme.cc#L32-L46) if level was less than or equal to 3, the arrays filled by `::detectorLevel` were not used - unless the calls in that function had side-effects

File: SimG4CMS/Forward/src/BscNumberingScheme.cc (hunk 3 above)
Comment (iarspider) [reply SimG4CMS/Forward/src/BscNumberingScheme.cc:55]: Will do

File: SimG4CMS/Forward/src/ZdcNumberingScheme.cc (hunk 4 above)
Comment (iarspider) [reply SimG4CMS/Forward/src/ZdcNumberingScheme.cc:110]: Will do

**Comment by iarspider** [comment]
> @iarspider , I fully agree that making temporary C-arrays in these classes is not necessary, in general, this PR is valid. However, these classes are used by sub-detectors, which in normal tests are not involved, so validation of results should be done with enough statistics or in the new code you should have another proof that the result is not changed. In a standard WF you likely will not see anything. A safe approach would be to do exactly the same thing as before but without C-arrays.

@civanch correct me if I'm wrong, but these classes are mostly just packing and unpacking integers. Or the tricky part is to somehow make sure `aStep->GetPreStepPoint()->GetTouchable();` iterates over all parts?

**Comment by civanch** [comment]
@iarspider , any hit should have a correct detector ID, which defined by numbering schema. It is critical in this PR to have exactly the same detID numbers as was before.

**Comment by civanch** [comment]
@iarspider , to me not clear why testing of this PR face problems while #41979 is working, let us close this.

**Pull request closed without merging**